docs/lbrdbd.1
.TH lbrdbd 1 "30 July 2015" "2.3.90" "lbrdbd man page"
.SH NAME
lbrdbd \- Rule database daemon for ldbox
.SH SYNOPSIS
.B lbrdbd
[options]
.SH DESCRIPTION
.I lbrdbd
initializes and manages a rule database inside every ldbox session.
.PP
Normally
.I lbrdbd
is started automatically when a session is created by
.I lb
and will be shut down when the session is
terminated, so there is usually no need to interact with this
daemon directly. However, under some conditions, it might be useful to 
specify options -S, -M or -F for lbrdbd. That can be done
with the "-x" option of lb.
.PP
.I lbrdbd
is responsible for all updates to the database; client processes (i.e.
all processes running inside the session) will use mmap(2) to map the
database image to memory, and will read data from there without
any kind of locking. If a client process needs to add or update data,
it will create a socket and send an RPC message to lbrdbd, which will
perform the update.
.PP
The database is used to hold several kinds of rules: During session
setup pathmapping rules and exec rules are written to it. Those won't
be modified during session lifetime. But rules related to the virtual
permissions (see lb(1)) might be added anytime, which might cause
the database file to grow.

.SH OPTIONS

.TP
\-d LEVEL
Enable debug messages.

.TP
\-f
foreground; does not fork (for debugging the daemon).

.TP
\-F MIN_CLIENT_SOCKET_FD
Set minimum client socket filedescriptor value.
Client processes which need to communicate with 
lbrdbd will create a socket, and then relocate the
file descriptor to a value which is at least
the value defined by this option.
Default is 279.

.TP
\-l FILE
Log messages to FILE. If FILE is "-", writes to stdout.
(note that
.I lb
redirects stdout of lbrdbd to $LDBOX_SESSION_DIR/lbrdbd.out
and uses option "-l -")

.TP
\-M MIN_MMAP_ADDR
Set minimum address for the memory mapped database.
Each client process will mmap(2) the database to be above
this address. Default is 0. This can be a 64-bit value,
but note that there can be both 64- and 32-bit client processes,
and the 32-bit processes will drop the most significant
32 bits, and only use the lower part of this value.

.TP
\-n
Don't start server; only initializes the rule database and exits (for debugging the daemon).

.TP
\-p FILE
write process ID to FILE.

.TP
\-s SESSION_DIR
set location of the session directory.

.TP
\-S SIZE
set maximum size of the memory mapped database.
This defines the amount of memory which will be reserved for
the database in
each client process running inside a ldbox session.
Default is 16 megabytes.

.SH DEBUGGING
A note for developers (of LB itself) about debugging:
The rule database file contains binary data. 
There is a debugging tool for printing out the contents,
distributed with the LB sources; see utils/lb-ruletree.c
if you ever need to dig into the low-level details.

.SH FILES
.P
.I $LDBOX_SESSION_DIR/*
.P

.SH SEE ALSO
.BR lb (1),
.BR lb-init (1),
.SH BUGS
No known bugs at this time.
.SH AUTHOR
.nf
Lauri T. Aarnio
.fi


lbrdbd/lbrdbd.c
/*
 * Copyright (C) 2011 Nokia Corporation.
 * Author: Lauri T. Aarnio
 *
 * Licensed under LGPL version 2.1, see top level LICENSE file for details.
*/

/* lbrdbd, the rule tree server:
 * The one and only process allowed to add data to the rule tree db
 * during an active LB session.
 *
 * Other processes may contact this server via a socket (AF_UNIX)
 * and ask the rule tree to be updated; reading from it is allowed
 * without locking, so care must be taken.
 *
 * This implementation assumes that writing 32-bit ints to the
 * memory-mapped file is an atomic operation.
*/

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdarg.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>

#include <sys/socket.h>
#include <sys/un.h>

#include <assert.h>

#include "lb_server.h"
#include "rule_tree_lua.h"

/* globals */

const char *progname = NULL;
char    *ldbox_session_dir = NULL;
char    *pid_file = NULL;


static void write_pid_to_file(pid_t s_pid, const char *pid_file)
{
	if (pid_file) {
		FILE *pidf = NULL;
		pidf = fopen(pid_file, "w");
		if (!pidf) {
			fprintf(stderr, "%s: Failed to open %s for writing\n",
				progname, pid_file);
		} else {
			fprintf(pidf, "%u\n", s_pid);
			fclose(pidf);
		}
	}
}

lua_State *lbrdbd_lua;

static void load_and_execute_lua_file(const char *filename)
{
	const char *errmsg;

	switch(luaL_loadfile(lbrdbd_lua, filename)) {
	case LUA_ERRFILE:
		fprintf(stderr, "Error loading %s\n", filename);
		exit(1);
	case LUA_ERRSYNTAX:
		errmsg = lua_tostring(lbrdbd_lua, -1);
		fprintf(stderr, "Syntax error in %s (%s)\n", filename, 
			(errmsg?errmsg:""));
		exit(1);
	case LUA_ERRMEM:
		fprintf(stderr, "Memory allocation error while "
				"loading %s\n", filename);
		exit(1);
	default:
		;
	}
	lua_call(lbrdbd_lua, 0, 0);
}

/* Lua calls this at panic: */
static int lb_lua_panic(lua_State *l)
{
	fprintf(stderr,
		"lbrdbd: Lua interpreter PANIC: unprotected error in call to Lua API (%s)\n",
		lua_tostring(l, -1));
#if 0
	lblog_init(); /* make sure the logger has been initialized */
#endif
	LB_LOG(LB_LOGLEVEL_ERROR,
		"Lua interpreter PANIC: unprotected error in call to Lua API (%s)\n",
		lua_tostring(l, -1));
	return 0;
}

static char *read_string_variable_from_lua(const char *name)
{
	char *result = NULL;

	if (name && *name) {
		lua_getglobal(lbrdbd_lua, name);
		result = (char *)lua_tostring(lbrdbd_lua, -1);
		if (result) {
			result = strdup(result);
		}
		lua_pop(lbrdbd_lua, 1);
		LB_LOG(LB_LOGLEVEL_DEBUG,
			"Lua variable %s = '%s', gettop=%d",
			name, (result ? result : "<NULL>"),
			lua_gettop(lbrdbd_lua));
	}
	return(result);
}

static void initialize_lua(void)
{
	char *main_lua_script = NULL;
	char *lua_if_version = NULL;

	if (asprintf(&main_lua_script, "%s/lua_scripts/init.lua",
	     ldbox_session_dir) < 0) {
		LB_LOG(LB_LOGLEVEL_ERROR,
			"%s: asprintf failed to allocate memory", __func__);
		return;
	}
		
	LB_LOG(LB_LOGLEVEL_INFO, "Loading '%s'", main_lua_script);

	lbrdbd_lua = luaL_newstate();
	lua_atpanic(lbrdbd_lua, lb_lua_panic);

	luaL_openlibs(lbrdbd_lua);
#if 0
	lua_bind_lb_functions(lbrdbd_lua); /* register our lb_ functions */
#endif
	lua_bind_ruletree_functions(lbrdbd_lua); /* register our ruletree_ functions */
	lua_bind_lblib_functions(lbrdbd_lua); /* register our lblib.* functions */

	load_and_execute_lua_file(main_lua_script);

	/* check Lua/C interface version. */
	lua_if_version = read_string_variable_from_lua("lbrdbd_lua_c_interface_version");
	if (!lua_if_version) {
		LB_LOG(LB_LOGLEVEL_ERROR, "FATAL ERROR: "
			"init.lua script didn't provide"
			" 'lbrdbd_lua_c_interface_version' identifier!");
		exit(1);
	}
	if (strcmp(lua_if_version, LBRDBD_LUA_C_INTERFACE_VERSION)) {
		LB_LOG(LB_LOGLEVEL_ERROR, "FATAL ERROR: "
			"init.lua script interface version mismatch:"
			" script has '%s', but '%s' was expected",
			lua_if_version, LBRDBD_LUA_C_INTERFACE_VERSION);
		exit(1);
	}
	free(lua_if_version);

	LB_LOG(LB_LOGLEVEL_INFO, "lua initialized.");
	LB_LOG(LB_LOGLEVEL_NOISE, "gettop=%d", lua_gettop(lbrdbd_lua));

	free(main_lua_script);
}

char *execute_init2_script(void)
{
	char	*init2_script = NULL;
	char	*result = NULL;

	if (asprintf(&init2_script, "%s/lua_scripts/init2.lua",
	     ldbox_session_dir) < 0) {
		LB_LOG(LB_LOGLEVEL_ERROR,
			"%s: asprintf failed to allocate memory", __func__);
		return(NULL);
	}
		
	LB_LOG(LB_LOGLEVEL_INFO, "Loading '%s'", init2_script);

	load_and_execute_lua_file(init2_script);

	/* get result. */
	result = read_string_variable_from_lua("init2_result");
	if (!result) {
		LB_LOG(LB_LOGLEVEL_ERROR, "init2 scripts didn't provide"
			" 'init2_result'!");
	}
	return(result);
}

static long long parse_num(const char *cp)
{
	long	l;
	char	*end = NULL;

	if (!cp || !*cp) {
		fprintf(stderr, "ERROR: A numeric argument is required");
		exit(1);
	}
	l = strtoll(cp, &end, 0);
	if (!end || (end == cp) || (*end != '\0')) {
		fprintf(stderr, "ERROR: %s is not a valid numeric argument", cp);
		exit(1);
	}
	return(l);
}

int main(int argc, char *argv[])
{
	int	opt;
	int	start_server = 1;
	int	backgroud_server = 1;
	char	*debug_level = NULL;
	char	*debug_file = NULL;
	char	*rule_tree_path = NULL;
	uint32_t max_size = 16*1024*1024; /* default 16MB */
	uint64_t min_mmap_addr = 0;
	int	min_client_socket_fd = 279;

	progname = argv[0];

	/* check data types: the rule tree DB uses plain types for
	 * certain data types; these assertions guarantee
	 * that LB startup fails if the OS is changed someday
	 * in the future, and data would not fit anumore to the DB...
	*/
	assert(sizeof(uint64_t) >= sizeof(dev_t));
	assert(sizeof(uint64_t) >= sizeof(ino_t));
	assert(sizeof(uint32_t) >= sizeof(uid_t));
	assert(sizeof(uint32_t) >= sizeof(gid_t));
	assert(sizeof(uint32_t) >= sizeof(mode_t));

	while ((opt = getopt(argc, argv, "d:l:s:p:nfS:M:F:")) != -1) {
		switch (opt) {
		case 'd':
			debug_level = strdup(optarg);
			break;
		case 'l':
			debug_file = strdup(optarg);
			break;
		case 's':
			ldbox_session_dir = strdup(optarg);
			break;
		case 'p':
			pid_file = strdup(optarg);
			break;
		case 'n': /* don't start server */
			start_server = 0;
			break;
		case 'f': /* foreground server (don't fork) */
			backgroud_server = 0;
			break;
		case 'S':
			max_size = parse_num(optarg);
			break;
		case 'M':
			min_mmap_addr = parse_num(optarg);
			break;
		case 'F':
			min_client_socket_fd = parse_num(optarg);
			break;
		default:
			fprintf(stderr, "Illegal option\n");
			exit(1);
		}
	}
	/* init logging before doing anything else,
	 * otherwise the logger will auto-initialize itself.
	 * if debug_level and/or debug_file is NULL, logger
	 * will read the values from env.vars. */
	lblog_init_level_logfile_format(debug_level, debug_file, NULL);

	if (!ldbox_session_dir) {
		fprintf(stderr, "ERROR: "
			"Option '-s session_dir' is mandatory.\n");
		exit(1);
	}

	/* ----- Initialization ----- */

	LB_LOG(LB_LOGLEVEL_DEBUG, "Initializing");

	if (asprintf(&rule_tree_path, "%s/RuleTree.bin", ldbox_session_dir) < 0) {
		exit(1);
	}

	if (create_ruletree_file(rule_tree_path,
		max_size, min_mmap_addr, min_client_socket_fd) < 0) {

		LB_LOG(LB_LOGLEVEL_ERROR, "Failed to create rule tree file (%s)",
			rule_tree_path);
		fprintf(stderr, "Failed to create rule DB!\n");
		exit(1);
	}
	LB_LOG(LB_LOGLEVEL_DEBUG, "Rule tree file opened & mapped to memory");

	initialize_lua();

	/* ----- Server ----- */
	if (start_server) {
		pid_t worker_pid;

		LB_LOG(LB_LOGLEVEL_DEBUG, "Initializing server");
		create_server_socket();

		/* Write PID to a file */
		if (backgroud_server) {
			if ((worker_pid = fork()) != 0) {
				/* parent */
				LB_LOG(LB_LOGLEVEL_DEBUG, "Parent process");
				write_pid_to_file(worker_pid, pid_file);
				return(0);
			}
		} else {
			write_pid_to_file(getpid(), pid_file);
		}
		
		/* enter the server loop. 
		 * ruletree_server() returns when the socket has been
		 * deleted and it is time to shut down. */
		ruletree_server();
	}
	return(0);
}

